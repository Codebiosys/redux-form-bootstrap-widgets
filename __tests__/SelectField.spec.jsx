/* eslint-disable import/first */

import React from 'react';
import { mount } from 'enzyme';
import Select from 'react-select';
import { SelectField } from 'index';

const fieldProps = {
  input: {
    name: 'fieldName',
    onChange: jest.fn(),
    onBlur: jest.fn(),
    onFocus: jest.fn(),
  },
  meta: {
    form: 'formName',
    pristine: false,
  },
  label: 'Radio Test',
  helpText: 'The help Text',
  options: [
    { label: 'Option 1', value: 'One' },
    { label: 'Option 2', value: 'Two' },
    { label: 'Option 3', value: 'Three' },
  ],
};

const selectedFieldProps = { ...fieldProps };
selectedFieldProps.input = { ...fieldProps.input };
selectedFieldProps.input.value = 'One';

const customOptionFieldProps = { ...fieldProps };
customOptionFieldProps.labelKey = 'foo';
customOptionFieldProps.valueKey = 'bar';
customOptionFieldProps.options = [
  { foo: 'Option 1', bar: 'One' },
  { foo: 'Option 2', bar: 'Two' },
  { foo: 'Option 3', bar: 'Three' }];


const multipleFieldProps = { ...fieldProps };
multipleFieldProps.multiple = true;

const multiValueFieldProps = { ...multipleFieldProps };
multiValueFieldProps.input = { ...multipleFieldProps.input };
multiValueFieldProps.input.value = ['One', 'Two'];

const asyncFieldProps = { ...fieldProps };
asyncFieldProps.options = input => Promise.resolve({ // eslint-disable-line no-unused-vars
  options: [
          { label: 'one', value: 1 },
          { label: 'two', value: 2 },
          { label: 'three', value: 3 },
  ],
});

describe('The Select Field', () => {
  let inputWrapper;
  let asyncWrapper;
  let multipleWrapper;

  beforeEach(() => {
    inputWrapper = mount(<SelectField {...fieldProps} />);
    asyncWrapper = mount(<SelectField {...asyncFieldProps} />);
    multipleWrapper = mount(<SelectField {...multipleFieldProps} />);
  });

  it('renders', () => {
    expect(inputWrapper).toMatchSnapshot();
  });


  it('has a label when there is a label', () => {
    expect(inputWrapper.find('ControlLabel').text()).toEqual(fieldProps.label);
  });

  it('renders with custom options', () => {
    inputWrapper = mount(<SelectField {...customOptionFieldProps} />);
    expect(inputWrapper).toMatchSnapshot();
  });

  it('renders with async options', () => {
    expect(asyncWrapper).toMatchSnapshot();
  });

  it('renders allowing multiple selections', () => {
    expect(multipleWrapper).toMatchSnapshot();
  });
  it('triggers the focus when focused', () => {
    inputWrapper.find('.Select-input input').simulate('focus');
    expect(fieldProps.input.onFocus).toHaveBeenCalledTimes(1);
  });

  it('triggers the async focus when async field is focused', () => {
    asyncWrapper.find('.Select-input input').simulate('focus');
    expect(asyncFieldProps.input.onFocus).toHaveBeenCalledTimes(1);
  });

  it('triggers the blur when blurred', () => {
    inputWrapper.find('.Select-input input').simulate('blur');
    expect(fieldProps.input.onBlur).toHaveBeenCalledTimes(1);
  });

  it('triggers the async blur when async field is blurred', () => {
    asyncWrapper.find('.Select-input input').simulate('blur');
    expect(asyncFieldProps.input.onBlur).toHaveBeenCalledTimes(1);
  });

  it('triggers the onChange when value is changed', () => {
    inputWrapper.find(Select).prop('onChange')({ label: 'Option 2', value: 'Two' });
    expect(fieldProps.input.onChange).toHaveBeenCalledTimes(1);
    expect(fieldProps.input.onChange).toHaveBeenCalledWith('Two');
  });

  it('triggers the onChange with null when value is missing', () => {
    inputWrapper.find(Select).prop('onChange')({ label: 'Option 4' });
    expect(fieldProps.input.onChange).toHaveBeenCalledWith(null);
  });

  it('triggers the onChange when value is changed for multiple entry', () => {
    multipleWrapper.find(Select).prop('onChange')({ label: 'Option 2', value: 'Two' });
    expect(multipleFieldProps.input.onChange).toHaveBeenCalledTimes(1);
    expect(multipleFieldProps.input.onChange).toHaveBeenCalledWith(['Two']);
  });

  it('adds to the value when  the onChange when value is changed for multiple entry', () => {
    multipleWrapper = mount(<SelectField {...multiValueFieldProps} />);
    multipleWrapper.find(Select).prop('onChange')([{ label: 'Option 1', value: 'One' }, { label: 'Option 3', value: 'Three' }]);
    expect(multipleFieldProps.input.onChange).toHaveBeenCalledTimes(1);
    expect(multipleFieldProps.input.onChange).toHaveBeenCalledWith(['One', 'Three']);
  });

  it('triggers the onChange with null when value is emtry', () => {
    multipleWrapper.find(Select).prop('onChange')([]);
    expect(multipleFieldProps.input.onChange).toHaveBeenCalledWith(null);
  });

  it('calls custom validator when the toogle is toggled', () => {
    const customValidator = jest.fn(() => ({ validationState: null, errorMessage: null }));
    const validatorProps = { ...fieldProps, customValidation: customValidator };
    const inputWrapperValidated = mount(<SelectField {...validatorProps} />);

    inputWrapperValidated.find(`input[name="${fieldProps.input.name}"]`).simulate('change', { target: { value: 'Three' } });
    expect(customValidator).toHaveBeenCalledTimes(1);
  });
});
