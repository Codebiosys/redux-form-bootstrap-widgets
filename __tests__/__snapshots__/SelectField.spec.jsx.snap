// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Select Field renders 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="fieldName"
>
  <Label
    label="Radio Test"
    required={false}
  />
  <Select
    arrowRenderer={[Function]}
    autoBlur={true}
    autosize={true}
    backspaceRemoves={true}
    backspaceToRemoveMessage="Press backspace to remove {label}"
    clearAllText="Clear all"
    clearRenderer={[Function]}
    clearValueText="Clear value"
    clearable={true}
    closeOnSelect={true}
    deleteRemoves={true}
    delimiter=","
    disabled={false}
    escapeClearsValue={true}
    filterOptions={[Function]}
    ignoreAccents={true}
    ignoreCase={true}
    inputProps={Object {}}
    isLoading={false}
    joinValues={true}
    labelKey="label"
    matchPos="any"
    matchProp="any"
    menuBuffer={0}
    menuRenderer={[Function]}
    multi={false}
    name="fieldName"
    noResultsText="No results found"
    onBlur={[Function]}
    onBlurResetsInput={true}
    onChange={[Function]}
    onCloseResetsInput={true}
    onFocus={[Function]}
    onSelectResetsInput={true}
    openOnClick={true}
    optionComponent={[Function]}
    options={
      Array [
        Object {
          "label": "Option 1",
          "value": "One",
        },
        Object {
          "label": "Option 2",
          "value": "Two",
        },
        Object {
          "label": "Option 3",
          "value": "Three",
        },
      ]
    }
    pageSize={5}
    placeholder="Select..."
    removeSelected={true}
    required={false}
    rtl={false}
    scrollMenuIntoView={true}
    searchable={true}
    simpleValue={false}
    tabSelectsValue={true}
    trimFilter={true}
    valueComponent={[Function]}
    valueKey="value"
  />
  <HelpBlock
    bsClass="help-block"
    style={
      Object {
        "minHeight": "3ex",
      }
    }
  />
</FormGroup>
`;

exports[`The Select Field renders allowing multiple selections 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="fieldName"
>
  <Label
    label="Radio Test"
    required={false}
  />
  <Select
    arrowRenderer={[Function]}
    autoBlur={true}
    autosize={true}
    backspaceRemoves={true}
    backspaceToRemoveMessage="Press backspace to remove {label}"
    clearAllText="Clear all"
    clearRenderer={[Function]}
    clearValueText="Clear value"
    clearable={true}
    closeOnSelect={true}
    deleteRemoves={true}
    delimiter=","
    disabled={false}
    escapeClearsValue={true}
    filterOptions={[Function]}
    ignoreAccents={true}
    ignoreCase={true}
    inputProps={Object {}}
    isLoading={false}
    joinValues={true}
    labelKey="label"
    matchPos="any"
    matchProp="any"
    menuBuffer={0}
    menuRenderer={[Function]}
    multi={true}
    name="fieldName"
    noResultsText="No results found"
    onBlur={[Function]}
    onBlurResetsInput={true}
    onChange={[Function]}
    onCloseResetsInput={true}
    onFocus={[Function]}
    onSelectResetsInput={true}
    openOnClick={true}
    optionComponent={[Function]}
    options={
      Array [
        Object {
          "label": "Option 1",
          "value": "One",
        },
        Object {
          "label": "Option 2",
          "value": "Two",
        },
        Object {
          "label": "Option 3",
          "value": "Three",
        },
      ]
    }
    pageSize={5}
    placeholder="Select..."
    removeSelected={true}
    required={false}
    rtl={false}
    scrollMenuIntoView={true}
    searchable={true}
    simpleValue={false}
    tabSelectsValue={true}
    trimFilter={true}
    valueComponent={[Function]}
    valueKey="value"
  />
  <HelpBlock
    bsClass="help-block"
    style={
      Object {
        "minHeight": "3ex",
      }
    }
  />
</FormGroup>
`;

exports[`The Select Field renders the help message with a break 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="fieldName"
  validationState={null}
>
  <Label
    label="Radio Test"
    required={false}
  />
  <Select
    arrowRenderer={[Function]}
    autoBlur={true}
    autosize={true}
    backspaceRemoves={true}
    backspaceToRemoveMessage="Press backspace to remove {label}"
    clearAllText="Clear all"
    clearRenderer={[Function]}
    clearValueText="Clear value"
    clearable={true}
    closeOnSelect={true}
    deleteRemoves={true}
    delimiter=","
    disabled={false}
    escapeClearsValue={true}
    filterOptions={[Function]}
    ignoreAccents={true}
    ignoreCase={true}
    inputProps={Object {}}
    isLoading={false}
    joinValues={true}
    labelKey="label"
    matchPos="any"
    matchProp="any"
    menuBuffer={0}
    menuRenderer={[Function]}
    multi={false}
    name="fieldName"
    noResultsText="No results found"
    onBlur={[Function]}
    onBlurResetsInput={true}
    onChange={[Function]}
    onCloseResetsInput={true}
    onFocus={[Function]}
    onSelectResetsInput={true}
    openOnClick={true}
    optionComponent={[Function]}
    options={
      Array [
        Object {
          "label": "Option 1",
          "value": "One",
        },
        Object {
          "label": "Option 2",
          "value": "Two",
        },
        Object {
          "label": "Option 3",
          "value": "Three",
        },
      ]
    }
    pageSize={5}
    placeholder="Select..."
    removeSelected={true}
    required={false}
    rtl={false}
    scrollMenuIntoView={true}
    searchable={true}
    simpleValue={false}
    tabSelectsValue={true}
    trimFilter={true}
    valueComponent={[Function]}
    valueKey="value"
  />
  <HelpBlock
    bsClass="help-block"
    style={
      Object {
        "minHeight": "6ex",
      }
    }
  >
    There was an error
    <br />
    The help Text
  </HelpBlock>
</FormGroup>
`;

exports[`The Select Field renders with async options 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="fieldName"
>
  <Label
    label="Radio Test"
    required={false}
  />
  <Async
    autoBlur={true}
    autoload={true}
    cache={Object {}}
    disabled={false}
    ignoreAccents={true}
    ignoreCase={true}
    inputProps={Object {}}
    joinValues={true}
    labelKey="label"
    loadOptions={[Function]}
    loadingPlaceholder="Loading..."
    multi={false}
    name="fieldName"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    options={Array []}
    searchPromptText="Type to search"
    valueKey="value"
  />
  <HelpBlock
    bsClass="help-block"
    style={
      Object {
        "minHeight": "3ex",
      }
    }
  />
</FormGroup>
`;

exports[`The Select Field renders with custom options 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="fieldName"
>
  <Label
    label="Radio Test"
    required={false}
  />
  <Select
    arrowRenderer={[Function]}
    autoBlur={true}
    autosize={true}
    backspaceRemoves={true}
    backspaceToRemoveMessage="Press backspace to remove {label}"
    clearAllText="Clear all"
    clearRenderer={[Function]}
    clearValueText="Clear value"
    clearable={true}
    closeOnSelect={true}
    deleteRemoves={true}
    delimiter=","
    disabled={false}
    escapeClearsValue={true}
    filterOptions={[Function]}
    ignoreAccents={true}
    ignoreCase={true}
    inputProps={Object {}}
    isLoading={false}
    joinValues={true}
    labelKey="foo"
    matchPos="any"
    matchProp="any"
    menuBuffer={0}
    menuRenderer={[Function]}
    multi={false}
    name="fieldName"
    noResultsText="No results found"
    onBlur={[Function]}
    onBlurResetsInput={true}
    onChange={[Function]}
    onCloseResetsInput={true}
    onFocus={[Function]}
    onSelectResetsInput={true}
    openOnClick={true}
    optionComponent={[Function]}
    options={
      Array [
        Object {
          "bar": "One",
          "foo": "Option 1",
        },
        Object {
          "bar": "Two",
          "foo": "Option 2",
        },
        Object {
          "bar": "Three",
          "foo": "Option 3",
        },
      ]
    }
    pageSize={5}
    placeholder="Select..."
    removeSelected={true}
    required={false}
    rtl={false}
    scrollMenuIntoView={true}
    searchable={true}
    simpleValue={false}
    tabSelectsValue={true}
    trimFilter={true}
    valueComponent={[Function]}
    valueKey="bar"
  />
  <HelpBlock
    bsClass="help-block"
    style={
      Object {
        "minHeight": "3ex",
      }
    }
  />
</FormGroup>
`;
